src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala

```diff
@@ -0,0 +1,286 @@
+   * Creates [[HadoopFsRelation]] based on the current relation.
+   *
+   * This is mainly used in conjunction with [[createLogicalRelation]].
+   */
+  override def createHadoopFsRelation(
+      location: FileIndex,
+      dataSchema: StructType,
+      options: Map[String, String]): HadoopFsRelation = plan.source match {
+    case _: IcebergSource =>
+      HadoopFsRelation(
+        location,
+        new StructType(),
+        dataSchema,
+        None,
+        new ParquetFileFormat,
+        options + IndexConstants.INDEX_RELATION_IDENTIFIER)(spark)
+  }
+
+  /**
+   * Creates [[LogicalRelation]] based on the current relation.
+   *
+   * This is mainly used to read the index files.
+   */
+  override def createLogicalRelation(
+      hadoopFsRelation: HadoopFsRelation,
+      newOutput: Seq[AttributeReference]): LogicalRelation = {
+    val updatedOutput =
+      newOutput.filter(attr => hadoopFsRelation.schema.fieldNames.contains(attr.name))
+    new LogicalRelation(hadoopFsRelation, updatedOutput, None, false)
+  }
+
+  private def getIcebergTable: Table = {
+    val options = new DataSourceOptions(plan.options.asJava)
+    val conf = spark.sessionState.newHadoopConf()
+    val path = options.get("path")
+    if (!path.isPresent) {
+      throw new IllegalArgumentException("Cannot open table: path is not set")
+    }
+    if (path.get.contains("/")) {
+      val tables = new HadoopTables(conf)
+      tables.load(path.get)
+    } else {
+      val hiveCatalog = HiveCatalogs.loadCatalog(conf)
+      val tableIdentifier = TableIdentifier.parse(path.get)
+      hiveCatalog.loadTable(tableIdentifier)
+    }
+  }
+
+  // TODO: need Iceberg's `tasks()` method to be public to get rid of this method
+  private def tasks(
+      table: Table,
+      options: DataSourceOptions,
+      pushedFilters: Seq[Filter],
+      requestedSchema: StructType): Seq[CombinedScanTask] = {
+    val caseSensitive = spark.conf.get("spark.sql.caseSensitive").toBoolean
+    val snapshotId = Try(options.get("snapshot-id").get().toLong).toOption
+    val asOfTimestamp = Try(options.get("as-of-timestamp").get().toLong).toOption
+    val startSnapshotId = Try(options.get("start-snapshot-id").get().toLong).toOption
+    val endSnapshotId = Try(options.get("end-snapshot-id").get().toLong).toOption
+    val splitSize = Try(options.get("split-size").get().toLong).toOption
+    val splitLookBack = Try(options.get("lookback").get().toInt).toOption
+    val splitOpenFileCost = Try(options.get("file-open-cost").get().toLong).toOption
+    val pushedExpressions = pushedFilters.map(SparkFilters.convert)
+    val lazySchema = if (requestedSchema != null) {
+      SparkSchemaUtil.prune(
+        table.schema,
+        requestedSchema,
+        pushedExpressions.reduceOption(Expressions.and).getOrElse(Expressions.alwaysTrue()),
+        caseSensitive)
+    } else {
+      table.schema
+    }
+
+    var scan: TableScan = table.newScan.caseSensitive(caseSensitive).project(lazySchema)
```

Comment:
nit: we don't specify types (unless required) for local variables.
```suggestion
    var scan = table.newScan.caseSensitive(caseSensitive).project(lazySchema)
```